### DISCUSSION TITLE
Modelling advice

### DISCUSSION THREAD

**Discussion by TimBailey-pnk** [statelyai/xstate#2562] [Q&A]
Hi, I'm struggling to determine best practice when modelling my scenario with xstate and I wonder if somebody could offer some advice.

I have a list of things, that can change either via a server request and/or some local filtering/ordering.  In this list of items, each item has some info lines which are added/deleted etc

I've modeled the list as my "listMachine" and I have a "itemMachine" for each item and then another "gridMachine" that represents the lines of info for each.

This model seems to be working fine, but now I'm struggling with basic features, 

The first being the ability to save each info change.  I want it to enter edit mode for a column in the info line, then save the value on enter/navigate.  But not save the value on escape key.  The way have it modelled now is saving on the `exit` of the `editingColumn`, but that then is preventing the cancel for cancelling.

Any thoughts on how to handle this?

**Comment by ChrisShank** [comment]
I think it would need some more detail on what you are trying to model to help with your specific use case, but from the sound of it I would check out the [TODOMVC example for xstate](https://codesandbox.io/s/xstate-todomvc-forked-hbnxs?file=/todoMachine.js) as there sounds to be a lot of overlap between functionality.

**Reply by TimBailey-pnk** [reply]
Hi, appreciate your time.  Thanks for the example I've had a look at it.

Some more details for my issue, I have the grid machine which represents a grid of infos for a item.  As usual each line is an info and each column is a property.

I have events that I send from my UI to represent the usual things that you can do when editing cells in a grid.  But my current problem resolves around the behaviour when saving followed by a secondary action.

So in the `cellSelected` state, I enter the `editingAmount` state when I press enter on the amount column. When I'm in this state I can edit the amount etc, but finally I want to save it.  The simplest case is to have the `KEYRETURN` event which saves the value and always moves back to the `cellSelected` state.  The behaviour that is giving me problems is where I want the `KEYARROWUP` to save the  value then change the current row and then move to `cellSelected` state.  I have a number of these arrow down, tab etc.  And then also I have the `KEYESCAPE` behavour which should not save but move to the `cellSelected` state.

Does that help to clarify what I'm trying to do?

Appreciate your thoughts.

Tim
